zcash_client_sqlite/src/lib.rs

```diff
@@ -910,12 +911,37 @@ impl<P: consensus::Parameters> WalletWrite for WalletDb<rusqlite::Connection, P>
                         }
 
                         // Ensure we have a Sapling checkpoint for each checkpointed Orchard block height
+                        // We skip all checkpoints below the minimum retained checkpoint in the
+                        // Sapling tree, because branches below this height may be pruned.
                         #[cfg(feature = "orchard")]
-                        for (height, checkpoint) in &missing_sapling_checkpoints {
-                            sapling_tree
-                                .store_mut()
-                                .add_checkpoint(*height, checkpoint.clone())
+                        {
+                            let min_checkpoint_height = sapling_tree
+                                .store()
+                                .min_checkpoint_id()
                                 .map_err(ShardTreeError::Storage)?;
+
+                            let add_checkpoint =
+                                |(height, checkpoint): &(BlockHeight, Checkpoint)| {
+                                    sapling_tree
+                                        .store_mut()
+                                        .add_checkpoint(*height, checkpoint.clone())
+                                        .map_err(ShardTreeError::Storage)
+                                };
+
+                            if let Some(h) = min_checkpoint_height {
+                                missing_sapling_checkpoints
+                                    .iter()
+                                    .filter(|(ch, _)| ch > &h)
+                                    .try_for_each(add_checkpoint)?;
+                            } else {
+                                let skip_len = missing_sapling_checkpoints
+                                    .len()
+                                    .saturating_sub(usize::try_from(PRUNING_DEPTH).unwrap());
```

Comment:
This skips checkpoints that are verifiably going to be pruned, but AFAICT it doesn't factor in any checkpoints we added from `from_state` or the scanned blocks. If that is intentional (i.e. we are still going to insert more checkpoints than will exist post-pruning, and are just calculating in this way for simplicity), add a comment.

Reply:
As noted, this whole clause can go away.